devito/ir/clusters/algorithms.py

```diff
@@ -314,3 +319,74 @@ def rule(size, e):
             processed.append(c.rebuild(exprs=exprs, ispace=ispace))
 
         return processed
+
+
+def normalize(clusters, **kwargs):
+    sregistry = kwargs['sregistry']
+
+    clusters = normalize_nested_indexeds(clusters, sregistry)
+    clusters = normalize_reductions(clusters, sregistry)
+
+    return clusters
+
+
+@cluster_pass(mode='all')
+def normalize_nested_indexeds(cluster, sregistry):
+    """
+    Recursively extract nested Indexeds in to temporaries.
+    """
+
+    def pull_indexeds(expr, subs, mapper, parent=None):
+        for i in retrieve_indexed(expr):
+            if i in mapper:
+                continue
+
+            for e in i.indices:
+                pull_indexeds(e, subs, mapper, parent=i)
+
+            if parent is not None:
+                # Nested Indexed, requires a temporary
+                k = i.xreplace(mapper)
+                v = Symbol(name=sregistry.make_name(), dtype=i.function.dtype)
+                subs[k] = v
+
+                # Update substitution status
+                mapper[i] = v
+
+    processed = []
+    for e in cluster.exprs:
+        subs = OrderedDict()
+        pull_indexeds(e, subs, {})
+
+        # Construct temporaries and apply substitution to `e`, in cascade
+        pe = e
```

Comment:
To avoid overriding the loop iteration variable

Reply:
Sure but your only changing the element not the iterator (the lenght/next/.. doesn't change) and since it's not used later and rebuilt it's just an extra (small arguably) unneeded overhead

Reply:
right, fixed